File: xsync_core/parallel.py
import os
import sys
import re
import shutil
import subprocess
import threading
import time
import tempfile

progress_pattern = re.compile(
    r"^\s*([\d,]+)\s+(\d+)%\s+([\d\.]+\w+/s)\s+[\d:]+\s+\(xfr#(\d+),\s+(?:to|ir|re)-chk=(\d+)/(\d+)\)"
)

def parse_line(line, job_id, progress_dict):
    match = progress_pattern.match(line)
    if match:
        percent = int(match.group(2))
        speed = match.group(3)
        xfer = int(match.group(4))
        total = int(match.group(6))
        
        progress_dict[job_id]["percent"] = percent
        progress_dict[job_id]["speed"] = speed
        progress_dict[job_id]["xfer"] = xfer
        progress_dict[job_id]["total"] = total
        progress_dict[job_id]["status"] = "Syncing"
    else:
        if not (line.startswith("sending ") or 
                line.startswith("receiving ") or 
                line.startswith("sent ") or 
                line.startswith("total size") or 
                line.startswith("speedup is") or
                line.startswith("building file list") or
                "bytes/sec" in line or
                "ir-chk" in line or
                "to-chk" in line):
            progress_dict[job_id]["current_file"] = line

def stdout_reader_thread(proc, job_id, progress_dict):
    buffer = b""
    while True:
        char = proc.stdout.read(1)
        if not char:
            break
        if char == b"\r" or char == b"\n":
            line = buffer.decode("utf-8", errors="ignore").strip()
            buffer = b""
            if line:
                parse_line(line, job_id, progress_dict)
        else:
            buffer += char
    
    progress_dict[job_id]["status"] = "Finished"
    progress_dict[job_id]["percent"] = 100
    progress_dict[job_id]["speed"] = "0 B/s"

def stderr_reader_thread(proc, job_id, errors_list):
    err = proc.stderr.read()
    if err:
        errors_list.append((job_id, err.decode("utf-8", errors="ignore")))

def run_parallel_sync(mode, local_dir, host, remote_dir, filter_file, delete, threads=4):
    ssh_control_path = f"/tmp/rsync-ctrl-{host}"
    ssh_cmd = f"ssh -o ControlMaster=auto -o ControlPersist=10m -o ControlPath={ssh_control_path} -c aes128-gcm@openssh.com -o Compression=no -o IPQoS=throughput"
    
    rsync_base = [
        "rsync",
        "-aW",
        "--partial",
        "-e", ssh_cmd
    ]

    # Source & Destination setup
    if mode == "push":
        src = local_dir.rstrip("/") + "/"
        dest = f"{host}:{remote_dir.rstrip('/')}/"
    else:
        src = f"{host}:{remote_dir.rstrip('/')}/"
        dest = local_dir.rstrip("/") + "/"

    print("  [INFO] Dang quet danh sach tep thay doi (Dry-run)...")
    dry_run_cmd = rsync_base + ["-v", "--dry-run"]
    if filter_file:
        dry_run_cmd.append(f"--filter=merge {filter_file}")
    if delete and mode == "push":
        dry_run_cmd.append("--delete")
    dry_run_cmd += [src, dest]

    proc = subprocess.Popen(dry_run_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    stdout, stderr = proc.communicate()
    
    if proc.returncode != 0:
        print(f"  [ERR] Dry-run that bai: {stderr.decode('utf-8', errors='ignore')}")
        return False

    lines = stdout.decode("utf-8", errors="ignore").splitlines()
    files_to_transfer = []
    files_to_delete = []

    for line in lines:
        line = line.strip()
        if not line:
            continue
        if line.startswith("deleting "):
            files_to_delete.append(line[len("deleting "):])
        elif (line.startswith("sending incremental") or 
              line.startswith("receiving incremental") or 
              line.startswith("sent ") or 
              line.startswith("total size") or 
              line.startswith("speedup is") or 
              line.startswith("created directory ") or
              "bytes/sec" in line):
            continue
        elif line.endswith("/"):
            continue
        else:
            files_to_transfer.append(line)

    print(f"  [INFO] Tim thay {len(files_to_transfer)} tep can truyen tai, {len(files_to_delete)} tep can xoa.")

    # 2. PERFORM DELETIONS FIRST IF ENABLED
    if delete and files_to_delete:
        print("  [INFO] Dang xoa cac tep thua tren dich...")
        if mode == "pull":
            # Local deletions
            for f in files_to_delete:
                local_path = os.path.join(local_dir, f)
                if os.path.isdir(local_path):
                    shutil.rmtree(local_path, ignore_errors=True)
                elif os.path.exists(local_path):
                    os.remove(local_path)
        else:
            # Remote deletions via SSH channel
            chunk_size = 300
            for i in range(0, len(files_to_delete), chunk_size):
                chunk = files_to_delete[i:i+chunk_size]
                quoted_paths = " ".join(f"'{os.path.join(remote_dir, f)}'" for f in chunk)
                ssh_del_cmd = [
                    "ssh",
                    "-o", f"ControlPath={ssh_control_path}",
                    host,
                    f"rm -rf {quoted_paths}"
                ]
                subprocess.run(ssh_del_cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        print("  [OK] Da hoan tat xoa tep thua.")

    if not files_to_transfer:
        print("  [OK] Khong co tep nao can truyen tai. Dong bo hoan tat!")
        return True

    # 3. SPLIT FILES INTO CHUNKS
    num_threads = min(threads, len(files_to_transfer))
    chunks = [[] for _ in range(num_threads)]
    for idx, f in enumerate(files_to_transfer):
        chunks[idx % num_threads].append(f)

    chunk_files = []
    for i in range(num_threads):
        temp_f = tempfile.NamedTemporaryFile(mode="w", suffix=f"_chunk_{i}.txt", delete=False)
        for path in chunks[i]:
            temp_f.write(path + "\n")
        temp_f.close()
        chunk_files.append(temp_f.name)

    # 4. LAUNCH PARALLEL RSYNC PROCESSES
    print(f"  [INFO] Dang khoi chay {num_threads} luong truyen tai song song...")
    processes = []
    progress_dict = {}
    reader_threads = []
    errors_list = []

    for i in range(num_threads):
        progress_dict[i] = {
            "percent": 0,
            "speed": "0 B/s",
            "xfer": 0,
            "total": len(chunks[i]),
            "current_file": "Dang ket noi...",
            "status": "Waiting"
        }
        
        cmd = rsync_base + [
            "-v",
            "--info=progress2",
            f"--files-from={chunk_files[i]}",
            src,
            dest
        ]
        
        proc = subprocess.Popen(
            cmd,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            bufsize=0
        )
        processes.append(proc)

        t_out = threading.Thread(target=stdout_reader_thread, args=(proc, i, progress_dict))
        t_out.daemon = True
        t_out.start()
        reader_threads.append(t_out)

        t_err = threading.Thread(target=stderr_reader_thread, args=(proc, i, errors_list))
        t_err.daemon = True
        t_err.start()
        reader_threads.append(t_err)

    # 5. LIVE PROGRESS DASHBOARD LOOP
    print("════════════════════════════════════════════════════════════════")
    print(" TIEN TRINH DONG BO PHAN LUONG (PARALLEL SYNC)")
    print("════════════════════════════════════════════════════════════════")
    for _ in range(num_threads):
        print()

    try:
        while True:
            all_finished = True
            for i in range(num_threads):
                if progress_dict[i]["status"] != "Finished":
                    all_finished = False
                    break

            sys.stdout.write(f"\033[{num_threads}F")
            
            for i in range(num_threads):
                job = progress_dict[i]
                percent = job["percent"]
                bar_width = 15
                filled = int(percent / 100 * bar_width)
                bar = "█" * filled + "░" * (bar_width - filled)
                
                file_info = job["current_file"]
                if len(file_info) > 30:
                    file_info = "..." + file_info[-27:]
                elif not file_info:
                    file_info = "-"
                
                if job["status"] == "Finished":
                    sys.stdout.write(f"\033[KLuong {i+1:02d}: [███████████████] 100% (Hoan thanh) - {job['total']}/{job['total']} tep\n")
                else:
                    sys.stdout.write(f"\033[KLuong {i+1:02d}: [{bar}] {percent:3d}% ({job['speed']}) - {file_info}\n")
            
            sys.stdout.flush()
            if all_finished:
                break
            time.sleep(0.2)
    except KeyboardInterrupt:
        print("\n  [WARN] Da dung truyen tai boi nguoi dung.")
        for proc in processes:
            if proc:
                proc.terminate()
        return False
    finally:
        for path in chunk_files:
            if path and os.path.exists(path):
                try:
                    os.remove(path)
                except OSError:
                    pass

    has_errors = False
    for i, proc in enumerate(processes):
        proc.wait()
        if proc.returncode != 0:
            has_errors = True

    if has_errors:
        print("════════════════════════════════════════════════════════════════")
        print("  [WARN] Co luong truyen tai that bai. Chi tiet loi:")
        for job_id, err_msg in errors_list:
            print(f"Luong {job_id+1}:")
            print(err_msg.strip())
        return False
    else:
        print("════════════════════════════════════════════════════════════════")
        print("  [OK] Dong bo tat ca cac luong hoan tat thanh cong!")
        return True

File: internal/parallel/parallel.go
package parallel

import (
	"bytes"
	"fmt"
	"io"
	"os"
	"os/exec"
	"os/signal"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"syscall"
	"time"
	"xsync/internal/tui"
)

var progressPattern = regexp.MustCompile(
	`^\s*([\d,]+)\s+(\d+)%\s+([\d\.]+\w+/s)\s+[\d:]+\s+\(xfr#(\d+),\s+(?:to|ir|re)-chk=(\d+)/(\d+)\)`,
)

type JobProgress struct {
	Percent     int
	Speed       string
	Xfer        int
	Total       int
	CurrentFile string
	Status      string
}

type ProgressTracker struct {
	mu   sync.RWMutex
	jobs []JobProgress
}

func (pt *ProgressTracker) Update(id int, fn func(*JobProgress)) {
	pt.mu.Lock()
	defer pt.mu.Unlock()
	if id >= 0 && id < len(pt.jobs) {
		fn(&pt.jobs[id])
	}
}

func (pt *ProgressTracker) Get(id int) JobProgress {
	pt.mu.RLock()
	defer pt.mu.RUnlock()
	if id >= 0 && id < len(pt.jobs) {
		return pt.jobs[id]
	}
	return JobProgress{}
}

func parseLine(line string, jobID int, tracker *ProgressTracker) {
	match := progressPattern.FindStringSubmatch(line)
	if len(match) >= 7 {
		pct, _ := strconv.Atoi(match[2])
		spd := match[3]
		xfr, _ := strconv.Atoi(match[4])
		tot, _ := strconv.Atoi(match[6])

		tracker.Update(jobID, func(j *JobProgress) {
			j.Percent = pct
			j.Speed = spd
			j.Xfer = xfr
			j.Total = tot
			j.Status = "Syncing"
		})
	} else {
		if !strings.HasPrefix(line, "sending ") &&
			!strings.HasPrefix(line, "receiving ") &&
			!strings.HasPrefix(line, "sent ") &&
			!strings.HasPrefix(line, "total size") &&
			!strings.HasPrefix(line, "speedup is") &&
			!strings.HasPrefix(line, "building file list") &&
			!strings.Contains(line, "bytes/sec") &&
			!strings.Contains(line, "ir-chk") &&
			!strings.Contains(line, "to-chk") {
			tracker.Update(jobID, func(j *JobProgress) {
				j.CurrentFile = line
			})
		}
	}
}

func stdoutReader(r io.Reader, jobID int, tracker *ProgressTracker) {
	buf := make([]byte, 1)
	var lineBuf bytes.Buffer

	for {
		n, err := r.Read(buf)
		if n > 0 {
			b := buf[0]
			if b == '\r' || b == '\n' {
				line := strings.TrimSpace(lineBuf.String())
				lineBuf.Reset()
				if line != "" {
					parseLine(line, jobID, tracker)
				}
			} else {
				lineBuf.WriteByte(b)
			}
		}
		if err != nil {
			break
		}
	}

	tracker.Update(jobID, func(j *JobProgress) {
		j.Status = "Finished"
		j.Percent = 100
		j.Speed = "0 B/s"
	})
}

func stderrReader(r io.Reader, jobID int, errorsList *[]string, mu *sync.Mutex) {
	data, _ := io.ReadAll(r)
	if len(data) > 0 {
		mu.Lock()
		*errorsList = append(*errorsList, fmt.Sprintf("Luong %02d:\n%s", jobID+1, strings.TrimSpace(string(data))))
		mu.Unlock()
	}
}

// RunParallelSync manages parallel rsync transfers.
func RunParallelSync(mode, localDir, host, remoteDir, filterFile string, deleteFiles bool, threads int) bool {
	sshControlPath := fmt.Sprintf("/tmp/rsync-ctrl-%s", host)
	sshCmd := fmt.Sprintf("ssh -o ControlMaster=auto -o ControlPersist=10m -o ControlPath=%s -c aes128-gcm@openssh.com -o Compression=no -o IPQoS=throughput", sshControlPath)

	rsyncBase := []string{
		"-aW",
		"--partial",
		"-e", sshCmd,
	}

	cleanLocal := strings.TrimRight(localDir, "/") + "/"
	cleanRemote := strings.TrimRight(remoteDir, "/") + "/"

	var src, dest string
	if mode == "push" {
		src = cleanLocal
		dest = fmt.Sprintf("%s:%s", host, cleanRemote)
	} else {
		src = fmt.Sprintf("%s:%s", host, cleanRemote)
		dest = cleanLocal
	}

	tui.PrintInfo("Dang quet danh sach tep thay doi (Dry-run)...")
	dryRunArgs := append([]string{}, rsyncBase...)
	dryRunArgs = append(dryRunArgs, "-v", "--dry-run")
	if filterFile != "" {
		dryRunArgs = append(dryRunArgs, fmt.Sprintf("--filter=merge %s", filterFile))
	}
	if deleteFiles && mode == "push" {
		dryRunArgs = append(dryRunArgs, "--delete")
	}
	dryRunArgs = append(dryRunArgs, src, dest)

	cmdDry := exec.Command("rsync", dryRunArgs...)
	outBytes, err := cmdDry.CombinedOutput()
	if err != nil {
		tui.PrintErr(fmt.Sprintf("Dry-run that bai: %s", string(outBytes)))
		return false
	}

	lines := strings.Split(string(outBytes), "\n")
	var filesToTransfer []string
	var filesToDelete []string

	for _, line := range lines {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		if strings.HasPrefix(line, "deleting ") {
			filesToDelete = append(filesToDelete, strings.TrimPrefix(line, "deleting "))
		} else if strings.HasPrefix(line, "sending incremental") ||
			strings.HasPrefix(line, "receiving incremental") ||
			strings.HasPrefix(line, "sent ") ||
			strings.HasPrefix(line, "total size") ||
			strings.HasPrefix(line, "speedup is") ||
			strings.HasPrefix(line, "created directory ") ||
			strings.Contains(line, "bytes/sec") {
			continue
		} else if strings.HasSuffix(line, "/") {
			continue
		} else {
			filesToTransfer = append(filesToTransfer, line)
		}
	}

	tui.PrintInfo(fmt.Sprintf("Tim thay %d tep can truyen tai, %d tep can xoa.", len(filesToTransfer), len(filesToDelete)))

	// 2. Perform deletions first if enabled
	if deleteFiles && len(filesToDelete) > 0 {
		tui.PrintInfo("Dang xoa cac tep thua tren dich...")
		if mode == "pull" {
			for _, f := range filesToDelete {
				target := filepath.Join(localDir, f)
				_ = os.RemoveAll(target)
			}
		} else {
			chunkSize := 300
			for i := 0; i < len(filesToDelete); i += chunkSize {
				end := i + chunkSize
				if end > len(filesToDelete) {
					end = len(filesToDelete)
				}
				chunk := filesToDelete[i:end]
				var quoted []string
				for _, f := range chunk {
					quoted = append(quoted, fmt.Sprintf("'%s'", filepath.Join(remoteDir, f)))
				}
				delCmdStr := fmt.Sprintf("rm -rf %s", strings.Join(quoted, " "))
				delCmd := exec.Command("ssh", "-o", fmt.Sprintf("ControlPath=%s", sshControlPath), host, delCmdStr)
				_ = delCmd.Run()
			}
		}
		tui.PrintOk("Da hoan tat xoa tep thua.")
	}

	if len(filesToTransfer) == 0 {
		tui.PrintOk("Khong co tep nao can truyen tai. Dong bo hoan tat!")
		return true
	}

	// 3. Split files into chunks
	numThreads := threads
	if len(filesToTransfer) < numThreads {
		numThreads = len(filesToTransfer)
	}

	chunks := make([][]string, numThreads)
	for i, f := range filesToTransfer {
		chunks[i%numThreads] = append(chunks[i%numThreads], f)
	}

	var chunkFiles []string
	defer func() {
		for _, cf := range chunkFiles {
			_ = os.Remove(cf)
		}
	}()

	for i := 0; i < numThreads; i++ {
		tmpF, err := os.CreateTemp("", fmt.Sprintf("*_chunk_%d.txt", i))
		if err != nil {
			tui.PrintErr(fmt.Sprintf("Khong the tao file chunk: %v", err))
			return false
		}
		for _, p := range chunks[i] {
			_, _ = tmpF.WriteString(p + "\n")
		}
		tmpF.Close()
		chunkFiles = append(chunkFiles, tmpF.Name())
	}

	// 4. Launch parallel processes
	tui.PrintInfo(fmt.Sprintf("Dang khoi chay %d luong truyen tai song song...", numThreads))

	tracker := &ProgressTracker{
		jobs: make([]JobProgress, numThreads),
	}
	for i := 0; i < numThreads; i++ {
		tracker.jobs[i] = JobProgress{
			Percent:     0,
			Speed:       "0 B/s",
			Xfer:        0,
			Total:       len(chunks[i]),
			CurrentFile: "Dang ket noi...",
			Status:      "Waiting",
		}
	}

	var procs []*exec.Cmd
	var errList []string
	var errMu sync.Mutex

	// Handle Ctrl+C to kill running rsync subprocesses
	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM)
	interrupted := false

	go func() {
		<-sigChan
		interrupted = true
		for _, p := range procs {
			if p != nil && p.Process != nil {
				_ = p.Process.Kill()
			}
		}
	}()

	for i := 0; i < numThreads; i++ {
		args := append([]string{}, rsyncBase...)
		args = append(args, "-v", "--info=progress2", fmt.Sprintf("--files-from=%s", chunkFiles[i]), src, dest)

		cmd := exec.Command("rsync", args...)
		stdoutPipe, err := cmd.StdoutPipe()
		if err != nil {
			tui.PrintErr(fmt.Sprintf("Loi stdout pipe luong %d: %v", i+1, err))
			return false
		}
		stderrPipe, err := cmd.StderrPipe()
		if err != nil {
			tui.PrintErr(fmt.Sprintf("Loi stderr pipe luong %d: %v", i+1, err))
			return false
		}

		if err := cmd.Start(); err != nil {
			tui.PrintErr(fmt.Sprintf("Loi khoi chay luong %d: %v", i+1, err))
			return false
		}
		procs = append(procs, cmd)

		go stdoutReader(stdoutPipe, i, tracker)
		go stderrReader(stderrPipe, i, &errList, &errMu)
	}

	// 5. Live progress dashboard
	fmt.Println("════════════════════════════════════════════════════════════════")
	fmt.Println(" TIEN TRINH DONG BO PHAN LUONG (PARALLEL SYNC)")
	fmt.Println("════════════════════════════════════════════════════════════════")
	for i := 0; i < numThreads; i++ {
		fmt.Println()
	}

	ticker := time.NewTicker(200 * time.Millisecond)
	defer ticker.Stop()

	for range ticker.C {
		if interrupted {
			fmt.Println("\n  [WARN] Da dung truyen tai boi nguoi dung.")
			return false
		}

		allFinished := true
		for i := 0; i < numThreads; i++ {
			if tracker.Get(i).Status != "Finished" {
				allFinished = false
				break
			}
		}

		fmt.Printf("\033[%dF", numThreads)
		for i := 0; i < numThreads; i++ {
			job := tracker.Get(i)
			barWidth := 15
			filled := int(float64(job.Percent) / 100.0 * float64(barWidth))
			if filled > barWidth {
				filled = barWidth
			}
			bar := strings.Repeat("█", filled) + strings.Repeat("░", barWidth-filled)

			fileInfo := job.CurrentFile
			if len(fileInfo) > 30 {
				fileInfo = "..." + fileInfo[len(fileInfo)-27:]
			} else if fileInfo == "" {
				fileInfo = "-"
			}

			if job.Status == "Finished" {
				fmt.Printf("\033[KLuong %02d: [███████████████] 100%% (Hoan thanh) - %d/%d tep\n", i+1, job.Total, job.Total)
			} else {
				fmt.Printf("\033[KLuong %02d: [%s] %3d%% (%s) - %s\n", i+1, bar, job.Percent, job.Speed, fileInfo)
			}
		}

		if allFinished {
			break
		}
	}

	// Wait for all subprocesses
	hasErrors := false
	for _, cmd := range procs {
		if err := cmd.Wait(); err != nil {
			hasErrors = true
		}
	}

	if hasErrors || len(errList) > 0 {
		fmt.Println("════════════════════════════════════════════════════════════════")
		tui.PrintWarn("Co luong truyen tai gap loi. Chi tiet:")
		for _, msg := range errList {
			fmt.Println(msg)
		}
		return false
	}

	fmt.Println("════════════════════════════════════════════════════════════════")
	tui.PrintOk("Dong bo tat ca cac luong hoan tat thanh cong!")
	return true
}
